examples/llm_compression/torch/qat_with_lora/main.py

```diff
@@ -0,0 +1,418 @@
+    parser.add_argument(
+        "--microbatch_size",
+        type=int,
+        default=8,
+        help="Size of each training microbatch. Gradients will be accumulated until the batch size is reached.",
+    )
+    return parser
+
+
+def main(argv) -> float:
+    """
+    Fine-tunes the specified model and returns the best validation similarity score.
+    """
+    parser = get_argument_parser()
+    args = parser.parse_args(argv)
+    assert torch.cuda.is_available()
+    set_seed(42)
+    device = "cuda"
+    torch_dtype = torch.bfloat16
+    compression_config = dict(
+        mode=CompressWeightsMode.INT4_ASYM, group_size=64, compression_format=CompressionFormat.FQ_LORA
+    )
+
+    # Configure output and log files.
+    output_dir = Path(args.output_dir)
+    tensorboard_dir = output_dir / "tb" / datetime.now().strftime("%D-%T")
+    last_dir = output_dir / "last"
+    best_dir = output_dir / "best"
+    for path in [output_dir, tensorboard_dir, last_dir, best_dir]:
+        if not args.resume:
+            shutil.rmtree(path, ignore_errors=True)
+        path.mkdir(exist_ok=True, parents=True)
+    wwb_ref_file = output_dir / "wwb_ref.csv"
+    ckpt_file = last_dir / "nncf_checkpoint.pth"
+    print(f"To visualize the loss and validation metrics, open Tensorboard using the logs from: {tensorboard_dir}")
+    tb = SummaryWriter(tensorboard_dir, "QAT with absorbable LoRA")
+
+    # Load original model and tokenizer.
+    model = AutoModelForCausalLM.from_pretrained(args.pretrained, torch_dtype=torch_dtype, device_map="auto")
+    tokenizer = AutoTokenizer.from_pretrained(args.pretrained)
+
+    # Use WhoWhatBench tool (WWB) is for validation during tuning. It estimates the similarity score between embedding
+    # computed by for data generated by two models, original floating-point one and optimized.
+    # TODO: (nlyalyus) Use original model for collecting reference, once the bug in WWB resolved.
+    wwb_ref_model = AutoModelForCausalLM.from_pretrained(args.pretrained, torch_dtype=torch_dtype, device_map="cpu")
+    save_wwb_ref(wwb_ref_model, tokenizer, wwb_ref_file)
+    del wwb_ref_model
+
+    # Prepare training data and pre-compute hiddens of teacher model for distillation loss.
+    train_loader = get_wikitext2(nsamples=args.nsamples, seqlen=args.seqlen, tokenizer=tokenizer, device=device)
+    orig_hiddens = calc_hiddens(model, train_loader)
+
+    # Create or load model to tune with Fake Quantizers and absorbable LoRA adapters.
+    example_input = get_model_input(train_loader[0])
+    if args.resume and ckpt_file.exists():
+        model = load_checkpoint(model, example_input, ckpt_file)
+    else:
+        model = compress_weights(model, dataset=Dataset([example_input]), **compression_config)
+        save_checkpoint(model, ckpt_file)
+    fq_lr = args.lr / 10
+    weight_decay = args.lr
+    param_to_train = set_trainable(model, lora_lr=args.lr, fq_lr=fq_lr)
```

Comment:
it would be great to redefine the `model.train()` call and move `set_trainable` inside it to simplify the UX. Right now, it looks overexposed. `lr` and `fq_lr` can be part of advanced params of `compress_weights`.

Reply:
Agree, that it would be great to hide `set_trainable` somehow. But not sure about overriding `model.train()`: since it would mean patching model's method.